### DISCUSSION TITLE
will multiple state instance be running concurrently if triggered multiple times?

### DISCUSSION THREAD

**Discussion by elim19** [serverlessworkflow/specification#745] [Q&A] [answered]
For a state like the following, if the OrderCreatedEvent triggered 3 times concurrently, will 3 state b instances be running concurrently ?

<img width="368" alt="image" src="https://user-images.githubusercontent.com/123845880/215307399-5c434cd6-5eb5-4988-979a-8118e70ce964.png">

**Comment by lsytj0413** [comment]
It's depend on the runtime implement, and basically speaking, the runtime will start 3 workflow instance. 

To make this idempotent, you can use `correlation` property. it was describe at https://github.com/serverlessworkflow/specification/blob/main/specification.md#event-definition

**Reply by elim19** [reply]
> https://github.com/serverlessworkflow/specification/blob/main/specification.md#event-definition

@lsytj0413 , as I understand, correlation is to define event correlation rules, could you give an example how to archive  idempotent in workflow instance?

**Comment by tsurdilo** [comment]
> and basically speaking, the runtime will start 3 workflow instance.

imho this is only true if this event state is the starting workflow state as the spec has it currently (and there is no event correlation defined)

**Comment by tsurdilo** [comment, marked as answer]
>  could you give an example how to archive idempotent in workflow instance?

im trying to understand this question, with SW you can only have one state active/running at any time, so the initial use case of multiple states running in parallel is not possible unless you use parallel state where each branch is a subflow for example

**Reply by elim19** [reply]
> im trying to understand this question, with SW you can only have one state active/running at any time, so the initial use case of multiple states running in parallel is not possible unless you use parallel state where each branch is a subflow for example

@tsurdilo , If this event state is the starting workflow state, can there be multiple state b active/running simultaneously?

**Reply by tsurdilo** [reply]
this goes back to what @lsytj0413 mentioned earlier, if you have a workflow definition with starting event state (and lets say we dont consider any event correlation rules) each event of type X would trigger invocation of a new workflow exec

maybe one thing we could do with correlation rules as they are currently is to have property that if set to true would mean correlate all events of type X (regardless of context attributes) to the execution that was triggered by the first one of that type..wdyt?

**Reply by elim19** [reply]
> this goes back to what @lsytj0413 mentioned earlier, if you have a workflow definition with starting event state (and lets say we dont consider any event correlation rules) each event of type X would trigger invocation of a new workflow exec
> 
> maybe one thing we could do with correlation rules as they are currently is to have property that if set to true would mean correlate all events of type X (regardless of context attributes) to the execution that was triggered by the first one of that type..wdyt?

@tsurdilo , after correlate all events of type X to the execution that was triggered by the first one of that type, what can you do about this correlation in SW? like limit the number the execution?

**Reply by tsurdilo** [reply]
no its more a business level mapping of events to workflow executions. think like "entity workflows" where an entity can be a patient, or user, or account, or whatever you do for your business
